jaxrs/src/main/java/com/facebook/airlift/jaxrs/thrift/ThriftMapper.java

```diff
@@ -0,0 +1,157 @@
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.MessageBodyReader;
+import javax.ws.rs.ext.MessageBodyWriter;
+import javax.ws.rs.ext.Provider;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+
+// This code is based on JacksonJsonProvider
+@Provider
+@Consumes("application/x-thrift")
+@Produces("application/x-thrift; qs=0.1")
+public class ThriftMapper
+        implements MessageBodyReader<Object>, MessageBodyWriter<Object>
+{
+    /**
+     * Looks like we need to worry about accidental
+     * data binding for types we shouldn't be handling. This is
+     * probably not a very good way to do it, but let's start by
+     * blacklisting things we are not to handle.
+     */
+    private static final ImmutableSet<Class<?>> IO_CLASSES = ImmutableSet.<Class<?>>builder()
+            .add(InputStream.class)
+            .add(java.io.Reader.class)
+            .add(OutputStream.class)
+            .add(java.io.Writer.class)
+            .add(byte[].class)
+            .add(char[].class)
+            .add(javax.ws.rs.core.StreamingOutput.class)
+            .add(Response.class)
+            .build();
+
+    public static final Logger log = Logger.get(ThriftMapper.class);
+
+    private final ThriftCodecManager thriftCodecManager;
+
+    @Inject
+    public ThriftMapper(ThriftCodecManager thriftCodecManager)
+    {
+        this.thriftCodecManager = thriftCodecManager;
+    }
+
+    @Override
+    public boolean isReadable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType)
+    {
+        return canReadOrWrite(type);
+    }
+
+    @Override
+    public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType)
+    {
+        return canReadOrWrite(type);
+    }
+
+    private static boolean canReadOrWrite(Class<?> type)
+    {
+        if (IO_CLASSES.contains(type)) {
+            return false;
+        }
+        for (Class<?> ioClass : IO_CLASSES) {
+            if (ioClass.isAssignableFrom(type)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public Object readFrom(Class<Object> type,
+            Type genericType,
+            Annotation[] annotations,
+            MediaType mediaType,
+            MultivaluedMap<String, String> httpHeaders,
+            InputStream inputStream)
+            throws IOException
+    {
+        try {
+            ThriftCodec<?> codec = thriftCodecManager.getCodec(type);
+            TProtocol tProtocol = new TFacebookCompactProtocol(new TInputStreamTransport(inputStream));
+            return codec.read(tProtocol);
+        }
+        catch (Exception e) {
+            // we want to return a 400 for bad Thrift but not for a real IO exception
+            if (e instanceof IOException && !(e instanceof EOFException)) {
```

Comment:
The IOException is likely to be wrapped into a `TTransportException`. Please handle this case